File: mParticle-Apple-SDK/Data Model/MPMessage.m
#import "MPMessage.h"
#import "MPSession.h"
#import "MPILogger.h"
#import "mParticle.h"
@import mParticle_Apple_SDK_Swift;

@interface MPMessage()
@property (nonatomic, strong) MPMessagePRIVATE *implementation;
@property (nonatomic, strong) NSData *messageData;
@end

@implementation MPMessage

- (instancetype)initWithSessionId:(NSNumber *)sessionId messageId:(int64_t)messageId UUID:(NSString *)uuid messageType:(NSString *)messageType messageData:(NSData *)messageData timestamp:(NSTimeInterval)timestamp uploadStatus:(MPUploadStatus)uploadStatus userId:(NSNumber *)userId dataPlanId:(NSString *)dataPlanId dataPlanVersion:(NSNumber *)dataPlanVersion {
    self = [super init];
    if (self) {
        _implementation = [[MPMessagePRIVATE alloc] initWithSessionId:sessionId
                                                            messageId:messageId
                                                                 UUID:uuid
                                                          messageType:messageType
                                                          messageData:messageData ?: [NSData data]
                                                            timestamp:timestamp
                                                         uploadStatus:(NSInteger)uploadStatus
                                                               userId:userId
                                                           dataPlanId:dataPlanId
                                                      dataPlanVersion:dataPlanVersion];
        _uuid = uuid;
    }
    return self;
}

+ (void)fixInvalidKeysInDictionary:(NSMutableDictionary *)messageDictionary messageInfo:(NSDictionary *)messageInfo {
    [MPMessagePRIVATE fixInvalidKeysInDictionary:messageDictionary messageInfo:messageInfo];
}

- (instancetype)initWithSession:(MPSession *)session messageType:(NSString *)messageType messageInfo:(NSDictionary *)messageInfo uploadStatus:(MPUploadStatus)uploadStatus UUID:(NSString *)uuid timestamp:(NSTimeInterval)timestamp userId:(NSNumber *)userId  dataPlanId:(NSString *)dataPlanId dataPlanVersion:(NSNumber *)dataPlanVersion {
    NSNumber *sessionId = nil;
    if (session) {
        sessionId = @(session.sessionId);
    }

    NSData *messageData = [MPMessagePRIVATE sanitizedJSONDataFromMessageInfo:messageInfo];
    return [self initWithSessionId:sessionId
                         messageId:0
                              UUID:uuid
                       messageType:messageType
                       messageData:messageData
                         timestamp:timestamp
                      uploadStatus:uploadStatus
                            userId:userId
                        dataPlanId:dataPlanId
                   dataPlanVersion:dataPlanVersion];
}

- (void)truncateMessageDataProperty:(NSString *)property toLength:(NSInteger)length {
    [self.implementation truncateMessageDataProperty:property toLength:length];
}

- (NSNumber *)sessionId { return self.implementation.sessionId; }
- (void)setSessionId:(NSNumber *)sessionId { self.implementation.sessionId = sessionId; }
- (int64_t)messageId { return self.implementation.messageId; }
- (void)setMessageId:(int64_t)messageId { self.implementation.messageId = messageId; }
- (NSString *)uuid { return self.implementation.uuid; }
- (void)setUuid:(NSString *)uuid { self.implementation.uuid = uuid; _uuid = uuid; }
- (NSString *)messageType { return self.implementation.messageType; }
- (NSData *)messageData { return self.implementation.messageData; }
- (void)setMessageData:(NSData *)messageData { self.implementation.messageData = messageData ?: [NSData data]; }
- (NSTimeInterval)timestamp { return self.implementation.timestamp; }
- (void)setTimestamp:(NSTimeInterval)timestamp { self.implementation.timestamp = timestamp; }
- (MPUploadStatus)uploadStatus { return (MPUploadStatus)self.implementation.uploadStatus; }
- (void)setUploadStatus:(MPUploadStatus)uploadStatus { self.implementation.uploadStatus = (NSInteger)uploadStatus; }
- (NSNumber *)userId { return self.implementation.userId; }
- (void)setUserId:(NSNumber *)userId { self.implementation.userId = userId; }
- (NSString *)dataPlanId { return self.implementation.dataPlanId; }
- (void)setDataPlanId:(NSString *)dataPlanId { self.implementation.dataPlanId = dataPlanId; }
- (NSNumber *)dataPlanVersion { return self.implementation.dataPlanVersion; }
- (void)setDataPlanVersion:(NSNumber *)dataPlanVersion { self.implementation.dataPlanVersion = dataPlanVersion; }
- (BOOL)shouldUploadEvent { return self.implementation.shouldUploadEvent; }
- (void)setShouldUploadEvent:(BOOL)shouldUploadEvent { self.implementation.shouldUploadEvent = shouldUploadEvent; }

- (NSString *)description {
    return [NSString stringWithFormat:@"Message\n Id: %lld\n UUID: %@\n Session: %@\n Type: %@\n timestamp: %.0f\n Data Plan: %@ %@\n Content: %@\n", self.messageId, self.uuid, self.sessionId, self.messageType, self.timestamp, self.dataPlanId, self.dataPlanVersion, [self serializedString]];
}

- (BOOL)isEqual:(MPMessage *)object {
    if (MPIsNull(object) || ![object isKindOfClass:[MPMessage class]]) {
        return NO;
    }
    return [self.implementation isEqualToMessage:object.implementation];
}

- (NSUInteger)hash {
    return (NSUInteger)self.implementation.hash;
}

- (id)copyWithZone:(NSZone *)zone {
    MPMessagePRIVATE *copyImpl = [self.implementation copyMessage];
    MPMessage *copyObject = [[MPMessage alloc] initWithSessionId:copyImpl.sessionId
                                                       messageId:copyImpl.messageId
                                                            UUID:copyImpl.uuid
                                                     messageType:copyImpl.messageType
                                                     messageData:copyImpl.messageData
                                                       timestamp:copyImpl.timestamp
                                                    uploadStatus:(MPUploadStatus)copyImpl.uploadStatus
                                                          userId:copyImpl.userId
                                                      dataPlanId:copyImpl.dataPlanId
                                                 dataPlanVersion:copyImpl.dataPlanVersion];
    copyObject.shouldUploadEvent = copyImpl.shouldUploadEvent;
    return copyObject;
}

- (void)encodeWithCoder:(NSCoder *)coder {
    [coder encodeObject:self.sessionId forKey:@"sessionId"];
    [coder encodeInt64:self.messageId forKey:@"messageId"];
    [coder encodeObject:self.uuid forKey:@"uuid"];
    [coder encodeObject:self.messageType forKey:@"messageType"];
    [coder encodeObject:self.messageData forKey:@"messageData"];
    [coder encodeDouble:self.timestamp forKey:@"timestamp"];
    [coder encodeInteger:self.uploadStatus forKey:@"uploadStatus"];
    [coder encodeInt64:self.userId.longLongValue forKey:@"mpid"];
    [coder encodeObject:self.dataPlanId forKey:@"dataPlanId"];
    [coder encodeObject:self.dataPlanVersion forKey:@"dataPlanVersion"];
    [coder encodeBool:self.shouldUploadEvent forKey:@"shouldUploadEvent"];
}

- (id)initWithCoder:(NSCoder *)coder {
    self = [self initWithSessionId:[coder decodeObjectOfClass:[NSNumber class] forKey:@"sessionId"]
                         messageId:[coder decodeInt64ForKey:@"messageId"]
                              UUID:[coder decodeObjectOfClass:[NSString class] forKey:@"uuid"]
                       messageType:[coder decodeObjectOfClass:[NSString class] forKey:@"messageType"]
                       messageData:[coder decodeObjectOfClass:[NSData class] forKey:@"messageData"]
                         timestamp:[coder decodeDoubleForKey:@"timestamp"]
                      uploadStatus:[coder decodeIntegerForKey:@"uploadStatus"]
                            userId:@([coder decodeInt64ForKey:@"mpid"])
                        dataPlanId:[coder decodeObjectOfClass:[NSString class] forKey:@"dataPlanId"]
                   dataPlanVersion:[coder decodeObjectOfClass:[NSNumber class] forKey:@"dataPlanVersion"]];
    if ([coder containsValueForKey:@"shouldUploadEvent"]) {
        self.shouldUploadEvent = [coder decodeBoolForKey:@"shouldUploadEvent"];
    }
    return self;
}

+ (BOOL)supportsSecureCoding {
    return YES;
}

- (NSDictionary *)dictionaryRepresentation {
    return [self.implementation dictionaryRepresentation];
}

- (NSString *)serializedString {
    return [self.implementation serializedString];
}

@end

File: mParticle-Apple-SDK-Swift/Sources/DataModel/MPMessage.swift
import Foundation

@objc(MPMessage)
public final class MPMessagePRIVATE: NSObject, NSCopying, NSSecureCoding {
    @objc public var sessionId: NSNumber?
    @objc public var messageId: Int64
    @objc public var uuid: String?
    @objc public var messageType: String?
    @objc public var messageData: Data?
    @objc public var timestamp: TimeInterval
    @objc public var uploadStatus: Int
    @objc public var userId: NSNumber
    @objc public var dataPlanId: String?
    @objc public var dataPlanVersion: NSNumber?
    @objc public var shouldUploadEvent = true

    @objc public static var supportsSecureCoding: Bool { true }

    @objc(initWithSessionId:messageId:UUID:messageType:messageData:timestamp:uploadStatus:userId:dataPlanId:dataPlanVersion:)
    public init(
        sessionId: NSNumber?,
        messageId: Int64,
        uuid: String?,
        messageType: String?,
        messageData: Data?,
        timestamp: TimeInterval,
        uploadStatus: Int,
        userId: NSNumber,
        dataPlanId: String?,
        dataPlanVersion: NSNumber?
    ) {
        self.sessionId = sessionId
        self.messageId = messageId
        self.uuid = uuid
        self.messageType = messageType
        self.messageData = messageData
        self.timestamp = timestamp
        self.uploadStatus = uploadStatus
        self.userId = userId
        self.dataPlanId = dataPlanId
        self.dataPlanVersion = dataPlanVersion
        super.init()
    }

    @objc(initWithSession:messageType:messageInfo:uploadStatus:UUID:timestamp:userId:dataPlanId:dataPlanVersion:)
    public convenience init(
        session: MPSessionPRIVATE?,
        messageType: String,
        messageInfo: NSDictionary,
        uploadStatus: Int,
        uuid: String,
        timestamp: TimeInterval,
        userId: NSNumber,
        dataPlanId: String?,
        dataPlanVersion: NSNumber?
    ) {
        self.init(
            sessionId: session.map { NSNumber(value: $0.sessionId) },
            messageId: 0,
            uuid: uuid,
            messageType: messageType,
            messageData: Self.sanitizedJSONData(from: messageInfo),
            timestamp: timestamp,
            uploadStatus: uploadStatus,
            userId: userId,
            dataPlanId: dataPlanId,
            dataPlanVersion: dataPlanVersion
        )
    }

    public required convenience init?(coder: NSCoder) {
        self.init(
            sessionId: coder.decodeObject(of: NSNumber.self, forKey: "sessionId"),
            messageId: coder.decodeInt64(forKey: "messageId"),
            uuid: coder.decodeObject(of: NSString.self, forKey: "uuid") as String?,
            messageType: coder.decodeObject(of: NSString.self, forKey: "messageType") as String?,
            messageData: coder.decodeObject(of: NSData.self, forKey: "messageData") as Data?,
            timestamp: coder.decodeDouble(forKey: "timestamp"),
            uploadStatus: coder.decodeInteger(forKey: "uploadStatus"),
            userId: NSNumber(value: coder.decodeInt64(forKey: "mpid")),
            dataPlanId: coder.decodeObject(of: NSString.self, forKey: "dataPlanId") as String?,
            dataPlanVersion: coder.decodeObject(of: NSNumber.self, forKey: "dataPlanVersion")
        )
        if coder.containsValue(forKey: "shouldUploadEvent") {
            shouldUploadEvent = coder.decodeBool(forKey: "shouldUploadEvent")
        }
    }

    @objc public func encode(with coder: NSCoder) {
        coder.encode(sessionId, forKey: "sessionId")
        coder.encode(messageId, forKey: "messageId")
        coder.encode(uuid, forKey: "uuid")
        coder.encode(messageType, forKey: "messageType")
        coder.encode(messageData, forKey: "messageData")
        coder.encode(timestamp, forKey: "timestamp")
        coder.encode(uploadStatus, forKey: "uploadStatus")
        coder.encode(userId.int64Value, forKey: "mpid")
        coder.encode(dataPlanId, forKey: "dataPlanId")
        coder.encode(dataPlanVersion, forKey: "dataPlanVersion")
        coder.encode(shouldUploadEvent, forKey: "shouldUploadEvent")
    }

    @objc(sanitizedJSONDataFromMessageInfo:)
    public static func sanitizedJSONData(from messageInfo: NSDictionary) -> Data? {
        let dictionary = NSMutableDictionary(dictionary: messageInfo)
        if !JSONSerialization.isValidJSONObject(messageInfo) {
            fixInvalidKeys(dictionary, messageInfo: messageInfo)
        }
        return try? JSONSerialization.data(withJSONObject: dictionary, options: [])
    }

    @objc(fixInvalidKeysInDictionary:messageInfo:)
    public static func fixInvalidKeys(_ messageDictionary: NSMutableDictionary, messageInfo: NSDictionary) {
        for key in messageInfo.allKeys {
            guard let key = key as? String else { continue }
            if let nested = messageInfo[key] as? NSDictionary {
                if !JSONSerialization.isValidJSONObject(nested) {
                    let mutableNested = (messageDictionary[key] as? NSDictionary)?.mutableCopy() as? NSMutableDictionary
                    if let mutableNested {
                        fixInvalidKeys(mutableNested, messageInfo: nested)
                    }
                    messageDictionary[key] = mutableNested
                }
            } else if let value = messageInfo[key] as? NSNumber {
                let doubleValue = value.doubleValue
                if doubleValue.isInfinite || doubleValue.isNaN {
                    NSLog("mParticle -> Invalid Message Data for key: %@", key)
                    NSLog("mParticle -> Value should not be infinite. Removing value from message data")
                    messageDictionary[key] = nil
                }
            }
        }
    }

    @objc(truncateMessageDataProperty:toLength:)
    public func truncateMessageDataProperty(_ property: String?, toLength length: Int) {
        guard let property, length >= 0, let payload = messageData else { return }
        guard let json = try? JSONSerialization.jsonObject(with: payload, options: []),
              let dictionary = (json as? NSDictionary)?.mutableCopy() as? NSMutableDictionary,
              let value = dictionary[property] as? String
        else {
            return
        }

        let bytes = Data(value.utf8)
        dictionary[property] = String(data: Data(bytes.prefix(min(bytes.count, length))), encoding: .utf8)
        messageData = try? JSONSerialization.data(withJSONObject: dictionary, options: [])
    }

    @objc public func dictionaryRepresentation() -> NSDictionary? {
        guard let messageData else { return nil }
        guard let dictionary = try? JSONSerialization.jsonObject(with: messageData, options: []) as? NSDictionary else {
            NSLog("mParticle -> Error serializing message.")
            return nil
        }
        return dictionary
    }

    @objc public func serializedString() -> String? {
        guard let messageData else { return nil }
        return String(data: messageData, encoding: .utf8)
    }

    @objc public func copy(with _: NSZone? = nil) -> Any {
        let copy = MPMessagePRIVATE(
            sessionId: sessionId,
            messageId: messageId,
            uuid: uuid,
            messageType: messageType,
            messageData: messageData,
            timestamp: timestamp,
            uploadStatus: uploadStatus,
            userId: userId,
            dataPlanId: dataPlanId,
            dataPlanVersion: dataPlanVersion
        )
        copy.shouldUploadEvent = shouldUploadEvent
        return copy
    }

    override public func isEqual(_ object: Any?) -> Bool {
        guard let other = object as? MPMessagePRIVATE else { return false }
        return isEqual(toMessage: other)
    }

    @objc(isEqualToMessage:)
    public func isEqual(toMessage other: MPMessagePRIVATE) -> Bool {
        sessionId == other.sessionId
            && messageId == other.messageId
            && timestamp == other.timestamp
            && stringsEqual(messageType, other.messageType)
            && dataEqual(messageData, other.messageData)
            && shouldUploadEvent == other.shouldUploadEvent
            && dataPlanId == other.dataPlanId
            && dataPlanVersion == other.dataPlanVersion
    }

    override public var hash: Int {
        var result = sessionId?.hashValue ?? 0
        result ^= dataPlanId?.hashValue ?? 0
        result ^= dataPlanVersion?.hashValue ?? 0
        result ^= Int(truncatingIfNeeded: messageId)
        result ^= Int(timestamp)
        result ^= messageType?.hashValue ?? 0
        result ^= messageData?.hashValue ?? 0
        result ^= shouldUploadEvent ? 1 : 0
        return result
    }

    override public var description: String {
        "Message\n Id: \(messageId)\n UUID: \(uuid ?? "(null)")\n Session: \(sessionId?.description ?? "(null)")"
            + "\n Type: \(messageType ?? "(null)")\n timestamp: \(String(format: "%.0f", timestamp))"
            + "\n Data Plan: \(dataPlanId ?? "(null)") \(dataPlanVersion?.description ?? "(null)")"
            + "\n Content: \(serializedString() ?? "(null)")\n"
    }

    private func stringsEqual(_ lhs: String?, _ rhs: String?) -> Bool {
        guard let lhs, let rhs else { return false }
        return lhs == rhs
    }

    private func dataEqual(_ lhs: Data?, _ rhs: Data?) -> Bool {
        guard let lhs, let rhs else { return false }
        return lhs == rhs
    }
}
